### PR TITLE
[ModuleInterface] Improvements to caching

### PR DESCRIPTION
**Pull request by graydon**
This picks up where https://github.com/apple/swift/pull/19518 left off, adding a new (very simple) dependency file type (called `.sid` since it's private anyways, and `.swiftinterfacedeps` seemed like it was getting a bit long) and improving the cache-key formation to match what @jrose-apple described over in https://forums.swift.org/t/thoughts-on-caching-compiled-parseable-interfaces/16746

Also factored out a common helper along the way that I was going to wind up adding another copy of if I didn't.

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

File: lib/Frontend/ParseableInterfaceSupport.cpp [hunk 1]
```diff
@@ -104,24 +143,58 @@ ParseableInterfaceModuleLoader::configureSubInvocationAndOutputPath(
   FEOpts.InputsAndOutputs.setMainAndSupplementaryOutputs({MainOut}, {SOPs});
 }
 
-// FIXME: this needs to be a more extensive up-to-date check.
+// Write thee world's simplest dependencies file: a version identifier on
```
Comment (harlanhaskins) [inline lib/Frontend/ParseableInterfaceSupport.cpp]: Nitpick: `the`
```suggestion
// Write the world's simplest dependencies file: a version identifier on
```

File: lib/Frontend/ParseableInterfaceSupport.cpp [hunk 2]
```diff
@@ -104,24 +143,58 @@ ParseableInterfaceModuleLoader::configureSubInvocationAndOutputPath(
...
+
+// Check that the output .swiftmodule file is at least as new as all the
+// dependencies it read when it was built last time.
 static bool
 swiftModuleIsUpToDate(clang::vfs::FileSystem &FS,
-                      StringRef InPath, StringRef OutPath) {
-  if (FS.exists(OutPath)) {
-    auto InStatus = FS.status(InPath);
+                      StringRef InPath, StringRef OutPath, StringRef DepPath) {
+  if (FS.exists(OutPath) && FS.exists(DepPath)) {
```
Comment (harlanhaskins) [inline lib/Frontend/ParseableInterfaceSupport.cpp]: Nit: Can you invert this check?

File: lib/Serialization/Serialization.cpp [hunk 3]
```diff
@@ -5158,19 +5159,6 @@ void Serializer::writeDocToStream(raw_ostream &os, ModuleOrSourceFile DC,
   S.writeToStream(os);
 }
 
-static inline bool
```
Comment (harlanhaskins) [inline lib/Serialization/Serialization.cpp]: Thank you for deduplicating this!

**Review by harlanhaskins — APPROVED** [review]
LGTM, modulo nitpicks

File: lib/Frontend/ParseableInterfaceSupport.cpp (hunk 1 above)
Comment (graydon) [reply lib/Frontend/ParseableInterfaceSupport.cpp]: Fixed.

File: lib/Frontend/ParseableInterfaceSupport.cpp (hunk 2 above)
Comment (graydon) [reply lib/Frontend/ParseableInterfaceSupport.cpp]: Done (redid control flow to be a bunch of early false returns, ending in success).

**Comment by graydon** [comment]
@swift-ci please test and merge

**Comment by graydon** [comment]
@swift-ci please test and merge

**Comment by jrose-apple** [comment]
Hm, I haven't looked over the source too closely, but why store this in a separate file, as opposed to part of the built swiftmodule?

**Pull request merged**